src/rust/inetstack/protocols/udp/peer.rs

```diff
@@ -286,13 +288,32 @@ impl<const N: usize> UdpPeer<N> {
     }
 
     /// Pops data from a socket.
-    pub fn do_pop(&self, qd: QDesc, size: Option<usize>) -> UdpPopFuture {
+    pub async fn do_pop(
+        qtable_ptr: Rc<RefCell<IoQueueTable>>,
+        qd: QDesc,
+        size: Option<usize>,
+    ) -> (QDesc, OperationResult) {
         #[cfg(feature = "profiler")]
         timer!("udp::pop");
-        let qtable: Ref<IoQueueTable> = self.qtable.borrow();
-        // Lookup associated receiver-side shared queue.
-        let queue: &UdpQueue = qtable.get::<UdpQueue>(&qd).expect("invalid queue descriptor");
-        UdpPopFuture::new(queue.get_recv_queue(), size)
+
+        // Get a reference to the shared queue, then drop the qtable reference.
+        let shared_queue: SharedQueue<SharedQueueSlot<DemiBuffer>> = {
+            let qtable: Ref<IoQueueTable> = qtable_ptr.borrow();
+            // Lookup associated receiver-side shared queue.
+            qtable
+                .get::<UdpQueue>(&qd)
+                .expect("queue should exist")
+                .get_recv_queue()
+                .clone()
+        };
```

Comment:
Instead of panicking, couldn't we log an error and return an `OperationResult::Failed`?